### PR TITLE
[Balloon] Added public documentation

### PR DESCRIPTION
**Pull request by georgepisaltu**
## Reason for This PR

The balloon device needs public documentation.

## Description of Changes

Added public documentation for the balloon device.

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [ ] All commits in this PR are signed (`git commit -s`).
- [ ] The reason for this PR is clearly provided (issue no. or explanation).
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] Any newly added `unsafe` code is properly documented.
- [ ] Any API changes are reflected in `firecracker/swagger.yaml`.
- [ ] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.

### CODE REVIEW

File: src/vmm/src/rpc_interface.rs [hunk 1]
```diff
@@ -1140,10 +1140,6 @@ mod tests {
             }),
             VmmActionError::OperationNotSupportedPreBoot,
         );
-        check_preboot_request_err(
```
Comment (raduiliescu) [inline src/vmm/src/rpc_interface.rs:1143]: Is this change intended?

Comment (georgepisaltu) [reply src/vmm/src/rpc_interface.rs:1143]: Yes, it was a duplicate check that slipped through in the "Merge to master" PR. You can find the same check in that file at line 1125.

File: docs/ballooning.md [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+
+curl --unix-socket $socket_location -i \
+    -X PUT 'http://localhost/balloon' \
+    -H 'Accept: application/json' \
+    -H 'Content-Type: application/json' \
+    -d "{
+        \"amount_mb\": $amount_mb, \
+        \"must_tell_host\": $must_tell_host, \
+        \"deflate_on_oom\": $deflate_on_oom, \
+        \"statistics_polling_interval_s\": $polling_interval \
```
Comment (dianpopa) [inline docs/ballooning.md:111]: it is "stats" not "statistics"

File: docs/ballooning.md [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+* not change the target size, which is set directly by the host
+* consume exactly as many pages as required to achieve the target size
+* correctly update the value of the actual size of the balloon seen by the host
+* not use pages that were previously inflated if they were not returned to the
+guest via a deflate operation (unless the `deflate_on_oom` flag was set and the
+guest is in an out of memory state)
+* provide correct statistics when available
+
+However, Firecracker does not and cannot introspect into the guest to check the
```
[the remark this answers is not in the document]
Comment (georgepisaltu) [reply docs/ballooning.md:48]: This paragraph is necessary as it is a core assumption in our threat model and the users need to know about this risk and protect against it.

The guest driver threat is the same across all our virtio devices, but it should nonetheless stated here. This is especially true as the balloon device introduces a new way to DoS the host in terms of memory usage. Unless clearly stated in documentation, an average user might use the balloon device thinking that once inflating the balloon to, let's say, half its VM size, the VM will never use more memory than half its original size. This is not the case if the driver is hijacked by a malicious guest. In the net device case, for example, the worst the guest can do is make the device operate to its max throughput (mitigated by our rate limiter), or cut their own net access. In the case of the balloon, a malicious guest could impact other VMs on the host.

File: docs/ballooning.md [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+of time in seconds at which the balloon statistics are updated.
+
+## Security disclaimer
+
+> [!IMPORTANT]
+> The balloon device is a paravirtualized virtio device that requires cooperation
+from a driver in the guest.
+
+In normal conditions, the balloon device will:
+
```
[the remark this answers is not in the document]
Comment (georgepisaltu) [reply docs/ballooning.md:39]: Removed.

File: docs/ballooning.md [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+it hits the target size. The following options are available:
+
+* `deflate_on_oom`: if this is set to `true`, then the kernel will deflate the
+balloon rather than entering an out-of-memory condition. Note that this does
+not mean that the kernel will deflate the balloon if it needs memory, but it
+would not enter an out-of-memory condition without it (i.e. it will not deflate
+the balloon to construct caches).
+* `must_tell_host`: if this is set to `true`, the kernel will wait for host
+confirmation before reclaiming memory from the host. This option is not useful
+in Firecracker's implementation of the balloon device, and can safely be set to
```
[the remark this answers is not in the document]
Comment (georgepisaltu) [reply docs/ballooning.md:26]: Added an explanation.

File: docs/ballooning.md [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+balloon (i.e. the memory it has allocated) is smaller than the target
+size, it continually tries to allocate new memory -- if it fails, it prints an
+error message, sleeps for some time, and then tries again. While the actual
+size of the balloon is larger than the target size, it will free memory until
+it hits the target size. The following options are available:
+
+* `deflate_on_oom`: if this is set to `true`, then the kernel will deflate the
+balloon rather than entering an out-of-memory condition. Note that this does
+not mean that the kernel will deflate the balloon if it needs memory, but it
+would not enter an out-of-memory condition without it (i.e. it will not deflate
```
[the remark this answers is not in the document]
Comment (georgepisaltu) [reply docs/ballooning.md:22]: Added an explanation.

File: docs/ballooning.md [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+A memory balloon device is a virtio device that can be used to reclaim guest
+memory. It does this by allocating memory in the guest, and then sending the
+addresses of that memory to the host; the host may then remove that memory at
+will. The device is configured through a number of options, and an integer,
+which represents the target size of the balloon, in MiB. The options cannot be
+changed during operation, but the target size can be changed.
+
+The behaviour of the balloon is the following: while the actual size of the
+balloon (i.e. the memory it has allocated) is smaller than the target
+size, it continually tries to allocate new memory -- if it fails, it prints an
```
[the remark this answers is not in the document]
Comment (georgepisaltu) [reply docs/ballooning.md:14]: Rephrased.

File: docs/ballooning.md [hunk 8]
```diff
@@ -0,0 +1,251 @@
+# Using the balloon device with Firecracker
+
+## What is the balloon device
+
+A memory balloon device is a virtio device that can be used to reclaim guest
```
[the remark this answers is not in the document]
Comment (georgepisaltu) [reply docs/ballooning.md:5]: Rephrased.

**Review by dianpopa — APPROVED** [review]

File: docs/ballooning.md [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+which represents the target size of the balloon, in MiB. The options cannot be
+changed during operation, but the target size can be changed.
+
+The behaviour of the balloon is the following: while the actual size of the
+balloon (i.e. the memory it has allocated) is smaller than the target
+size, it continually tries to allocate new memory -- if it fails, it prints an
+error message (`Out of puff! Can't get %d pages`), sleeps for 0.2 seconds, and
+then tries again. While the actual
+size of the balloon is larger than the target size, it will free memory until
+it hits the target size. The following options are available:
```
Comment (raduiliescu) [inline docs/ballooning.md:19]: Are the options related to how balloon is freeing memory? Otherwise please put the following options... in a new paragraph. Also rephrase to the balloon device can be configured with the following options

There is a line break nit above while the actual....

File: docs/ballooning.md [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+memory addresses, the memory is zeroed. Furthermore, the guest memory is
+`mmap`ped with the `MAP_PRIVATE` and `MAP_ANONYMOUS` flags, which ensure that
+even if a Firecracker yields some information through an inflate and that
+same physical page containing the information is mapped onto another
+Firecracker process, reads on that address space will see zeroes.
+
+## Prerequisites
+
+To support memory ballooning, you must use a kernel that has the memory
+ballooning driver installed (on Linux 4.14.7, the relevant settings are
```
Comment (gc-plp) [inline docs/ballooning.md:89]: Why specifically `4.14.7`?

File: docs/ballooning.md [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+desired: `num_pages` represents the target size of the balloon, and
+`must_tell_host`, `deflate_on_oom` and `stats_polling_interval_s`
+represent the options mentioned before.
+
+To install the balloon via the JSON config file, insert the following JSON
+object into your configuration file:
+
+```
+"balloon": {
+    "amount_mb": $amount_mb,
```
Comment (gc-plp) [inline docs/ballooning.md:133]: The example here is a bit strange as we're giving a JSON file as an example with variables as values.

We could replace them with predefined values, since this is just an example after all.

File: docs/ballooning.md (hunk 10 above)
Comment (georgepisaltu) [reply docs/ballooning.md:89]: I changed it to `4.14.193` since that is the config that we use in our tests.

File: docs/ballooning.md (hunk 11 above)
Comment (georgepisaltu) [reply docs/ballooning.md:133]: Changed.

**Review by georgepisaltu — COMMENTED** [review]
@raduiliescu @dianpopa @gc-plp Updated the doc, PTAL!

**Review by dianpopa — APPROVED** [review]

**Review by raduiliescu — DISMISSED** [review]

**Review by dianpopa — APPROVED** [review]

**Review by gc-plp — APPROVED** [review]
